Modify rule S2091(Java): Add Educational content (APPSEC-99) (#1198)
SonarSource/rspec: rules/S2091/java/how-to-fix-it/java-se.adoc

=== How to fix it in Java SE

> The following code is vulnerable to Xpath injections because untrusted data is

Reviewer: I have been doing this as well on my rules but maybe we should start specifying since the 'following code' expression also captures de compliant solution.
```suggestion
The following non-compliant code is vulnerable to Xpath injections because untrusted data is
```

Author: Hmm, don't worry about this IMHO. This is going to be displayed as a diff so I think we can keep this pattern

> injected in an Xpath query without prior validation.

Reviewer: ```suggestion
concatenated to an Xpath query without prior validation.
```

[cols="a"]
|===
h| Non-compliant code example
|
[source,java]
----
public boolean authenticate(HttpServletRequest req, XPath xpath, Document doc) throws XPathExpressionException { 
  String user = request.getParameter("user"); 
  String pass = request.getParameter("pass");

  String expression = "/users/user[@name='" + user + "' and @pass='" + pass + "']";

  return (boolean)xpath.evaluate(expression, doc, XPathConstants.BOOLEAN); // Noncompliant
}
----
h| Compliant solution
|
[source,java]
----
public boolean authenticate(HttpServletRequest req, XPath xpath, Document doc) throws XPathExpressionException { 
  String user = request.getParameter("user"); 
  String pass = request.getParameter("pass");

  String expression = "/users/user[@name=$user and @pass=$pass]";

  xpath.setXPathVariableResolver(v -> {
    switch (v.getLocalPart()) {
      case "user":
        return user;
      case "pass":
        return pass;
      default:
        throw new IllegalArgumentException();
    }
  });

  return (boolean)xpath.evaluate(expression, doc, XPathConstants.BOOLEAN);
}

----
|===

=== How does this work?

include::../../common/fix/validation.adoc[]

> In the example, a sanitization mechanism is applied to an XPath Factory as a

Reviewer: ```suggestion
In the compliant solution example, a sanitization mechanism is applied to an XPath Factory as a
```

Author: same as the "non-compliant" solution. IMHO we can trust the diff view to make it clear.

resolver to resolve the untrusted data to the request in a safe way, similarly
to parametrized SQL queries.

> 

Reviewer: Can you call it 'parameterized XPath queries instead on 'a sanitization mechanism' like they do in [OWASP XPATH_Injection](https://owasp.org/www-community/attacks/XPATH_Injection) .

Author: ```suggestion
In the example, a parameterized XPath query is created as an XPath Factory
resolver to insert untrusted data to the request in a safe way, similarly to
parametrized SQL queries.

```
